url6dc6a3bff8_JasonxJiang_PokemonEvolutionaryAlgorithm_tgz-pJ8-ee_ea_pkg1_EEEA1J8/src/ee/ea/pkg1/pokemon/GiratinaO.java

```diff
@@ -0,0 +1,197 @@
+            IVSPEED = 31;
+            EVHP = 6;
+            EVATK = 126;
+            EVDEF = 0;
+            EVSPA = 126;
+            EVSDF = 0;
+            EVSPEED = 252;
+            statBoostATK = 1;
+            statBoostDEF = 1;
+            statBoostSPA = 1;
+            statBoostSDF = 1;
+            statBoostSPEED = 1;
+            level = 100;
+            finalHp = (int) ((((IVHP + (2 * baseHealth) + (EVHP / 4)) * level) / 100) + 10);
+            finalAtk = (int) (((((IVATK + (2 * baseAttack) + (EVATK / 4)) * level) / 100) + 5) * natureAtk);
+            finalDef = (int) (((((IVDEF + (2 * baseDefense) + (EVDEF / 4)) * level) / 100) + 5) * natureDef);
+            finalSpa = (int) (((((IVSPA + (2 * baseSpecialAttack) + (EVSPA / 4)) * level) / 100) + 5) * natureSpa);
+            finalSdf = (int) (((((IVSPD + (2 * baseSpecialDefense) + (EVSDF / 4)) * level) / 100) + 5) * natureSdf);
+            finalSpeed = (int) (((((IVSPEED + (2 * baseSpeed) + (EVSPEED / 4)) * level) / 100) + 5) * natureSpeed);
+            name = "Giratina-O";
+            primary = new Dragon();
+            secondary = new Ghost();
+            moveLibrary = new HashMap<String, Move>();
+            Move m1 = new DracoMeteor();
+            Move m2 = new ShadowSneak();
+            Move m3 = new HiddenPowerFire();
+            Move m4 = new Earthquake();
+            Move m5 = new AuraSphere();
+            Move m6 = new DragonPulse();
+            Move m7 = new ShadowBall();
+            Move m8 = new Thunder();
+            moveLibrary.put("draco meteor", m1);
+            moveLibrary.put("shadow sneak", m2);
+            moveLibrary.put("hidden power fire", m3);
+            moveLibrary.put("earthquake", m4);
+            moveLibrary.put("aura sphere", m5);
+            moveLibrary.put("dragonpulse", m6);
+            moveLibrary.put("shadow ball", m7);
+            moveLibrary.put("thunder", m8);
+            potentials = new ArrayList<String>();
+            potentials.add("draco meteor");
+            potentials.add("shadow sneak");
+            potentials.add("hidden power fire");
+            potentials.add("earthquake");
+            potentials.add("aura sphere");
+            potentials.add("dragon pulse");
+            potentials.add("shadow ball");
+            potentials.add("thunder");
+            circumvent = new ArrayList<String>();
+            battleReady = new HashMap<String, Move>();
+            for (int i = 0; i < 4; i++) {
+                int randomIndex = (int) (Math.random() * potentials.size());
+                String key = potentials.get(randomIndex);
+                while (circumvent.contains(key)) {
+                    int nextRandom = (int) (Math.random() * potentials.size());
+                    key = potentials.get(nextRandom);
+                }
+                battleReady.put(key, moveLibrary.get(key));
+                circumvent.add(key);
+                //potentials.remove(randomIndex);
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 98:108, 183:193.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.